doc/website/advanced/how-optional-and-error-values-are-returned-in-iceoryx.md

```diff
@@ -0,0 +1,104 @@
+# How optional and error values are returned in iceoryx
+
+Many parts of the iceoryx C++ API follow a functional programming approach and allow the user to specify functions 
+which handle the possible cases, e.g. what should happen when data is received.
+
+This is very flexible but requires using the monadic types ``cxx::expected`` and ``cxx::optional``, which we
+introduce in the following sections.
+
+### Optional
+
+The type ``iox::cxx::optional<T>`` is used to indicate that there may or may not be a value of a specific type ``T``
+available. This is essentially the maybe [monad](https://en.wikipedia.org/wiki/Monad_(functional_programming)) in 
+functional programming. Assuming we have some optional (usually the result of some computation)
+
+```cpp
+optional<int> result = someComputation();
+```
+
+we can check for its value using
+```cpp
+if(result.has_value())
+{
+    auto value = result.value();
+    // do something with the value
+}
+else
+{
+    // handle the case that there is no value
+}
+```
+A shorthand to get the value is
+```cpp
+auto value = *result;
+```
+
+!!! attention
+    Accessing the value if there is no value is undefined behavior, so it must be checked beforehand.
+
+We can achieve the same with the functional approach by providing a function for both cases.
+
+```cpp
+result.and_then([](int& value) { /*do something with the value*/ })
+    .or_else([]() { /*handle the case that there is no value*/ });
+```
+Notice that we get the value by reference, so if a copy is desired it has to be created explicitly in the 
+[lambda](https://en.wikipedia.org/wiki/Anonymous_function#C++_(since_C++11)) or function we pass.
+
+The optional can be initialized from a value directly
+```cpp
+optional<int> result = 73;
+result = 37;
+```
+If it is default initialized it is automatically set to its null value of type ``iox::cxx::nullopt_t``. This can be also 
+done directly by using the constant ``iox::cxx::nullopt``
+
+```cpp
+result = iox::cxx::nullopt;
+```
+
+For a complete list of available functions see 
+[``optional.hpp``](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_utils/include/iceoryx_utils/cxx/optional.hpp).
+
+### Expected
+``iox::cxx::expected<T, E>`` generalizes ``iox::cxx::optional`` by admitting a value of another type ``E`` instead of
+no value at all, i.e. it contains either a value of type ``T`` or ``E``. In this way, ``expected`` is a special case of 
+the either monad. It is usually used to pass a value of type ``T`` or an error that may have occurred, i.e. ``E`` is the
```

Comment:
i would set "either monad" in quotes and maybe a link can be added, like the other monads.

Reply:
added quotation marks but not the link since it would be the same